modules/javafx.swing/src/main/java/javafx/embed/swing/JFXPanel.java

```diff
@@ -407,14 +446,17 @@ private void sendMouseEventToFX(MouseEvent e) {
         if (e.getID() == MouseEvent.MOUSE_PRESSED || e.getID() == MouseEvent.MOUSE_RELEASED) {
             popupTrigger = e.isPopupTrigger();
         }
+        Dimension2D onScreen = getSwingToFxPixel(getGraphicsConfiguration(), e.getXOnScreen(), e.getYOnScreen());
+        int fxXOnScreen = (int) onScreen.getWidth();
+        int fxYOnScreen = (int) onScreen.getHeight();
```

Comment:
question: should it be (int) or Math.round()?
also, coordinates can be negative - will it work then?

Reply:
I believe these values should be int in the first place. Otherwise, we have issues in JavaFX where we process them as ints.
About negative coordinates: yes, that works. Is the question about using Dimension2D (width/height) for something that is actually a Point2D? We can use a Point2D here as well, but that class seems to have a bit more overhead than Dimension2D for this goal -- but I'm open to change it to Point2D.

Reply:
No, what I meant is this:
```
 System.out.println("(int)-1.9999=" + (int)-1.9999 + " (int)Math.round(-1.9999)=" + (int)(Math.round(-1.9999)));
prints
(int)-1.9999=-1 (int)Math.round(-1.9999)=-2
```

i.e. typecast to int is equivalent to Math.ceil for negative values and Math.floor for positive.  We probably should use Math.round().

This code should be better tested on Windows with
a) second monitor positioned to the left of the main one, making its coordinates negative
b) using fractional HiDPI scaling which is only available on Windows it seems

And yes, I did not see Dimension2D - we probably should remove confusion and use Point2D (why does it have more overhead?)

Reply:
Depending on the desired result, we use `Math.floor`, `Math.ceil`, or `Math.round` -- it's important to pick the right one.

But yes, using one of those is (almost always) better than just casting to `(int)` if the value might be negative.

Reply:
I think in this case, `Math.floor()` is what you want. I wonder, though, whether it needs to be converted to an `int` at all? It looks like everywhere it is used, it can be kept as a double.

Btw, I almost wrote a comment asking a question as to whether you were using the values of x, y, width, and height correctly, because I got confused that you were using a Dimension to hold an x,y point. If you decide to keep it as a Dimension, I might suggest adding a clarifying comment.

Reply:
We need the int values, as they are used to invoke `mouseEvent`, `scrollEvent` and `menuEvent` on `com.sun.javafx.embed.EmbeddedSceneInterface` . 
For some reason, `scrollEvent` takes doubles, but `mouseEvent` and `menuEvent` take ints.
I'll use `Math.floor()` instead of the `int` cast. The key thing here is that the strategy for the mouse/scroll/menu events should be the same as the strategy for the window location changes.

Reply:
> And yes, I did not see Dimension2D - we probably should remove confusion and use Point2D (why does it have more overhead?)

Point2D provides much more functionality -- but has 2 fields only. So it should not be an overhead at all, hence I'll switch to Point2D. Thanks for Pointing that out :)

Reply:
I suspect Math.floor() would be incorrect - we should use Math.round() for coordinates and Math.floor() for sizes (sizes will always be positive, I hope, and the rounded size will be slightly smaller than the original value).

But for coordinates we should use (int)Math.round()

Reply:
I'll change it to Math.round, although I doubt it doesn't improve anything (but it won't make it worse either).
The reason we need the transformation on those coordinates is to position new windows, *relative* against the location of the root window that we are given by Swing. Hence, if the operation applied to the window location coordinates (in sendMoveEventToFX) is the same as the operation applied to the mouse point (in sendMouseEventToFX), we are good. 
A bigger problem imho is that mouseEvents, scrollEvents and menuEvents are dealing with different precisions in the EmbeddedSceneInterface -- but that is a different problem.

Reply:
I also expect that it won't make much of a difference here, but Andy is right that, for snapping positions, `Math.round` is preferred. That's what layout panes do in `Region::snapPositionX/Y`